pkg/datapath/linux/routing/migrate.go

```diff
@@ -0,0 +1,242 @@
+
+	return linux_defaults.RouteTableInterfacesOffset + eni.Number, nil
+}
+
+// migrateOnUpgrade migrates the ENI routing table for the upgrade path. It
+// accepts the old table ID for the existing rules to migrate. `new` is the new
+// table ID that the rules should be referring to.
+func migrateOnUpgrade(old, new int) error {
+	if old == new {
+		// Nothing to do.
+		return nil
+	}
+
+	if err := replaceDefaultRoutes(old, new); err != nil {
+		return err
+	}
+	if err := createNewRules(
+		linux_defaults.RulePriorityEgress,
+		old,
+		linux_defaults.RulePriorityEgressv2,
+		new,
+	); err != nil {
+		return err
+	}
+	if err := deleteOldRules(linux_defaults.RulePriorityEgress); err != nil {
+		return err
+	}
+	if err := deleteOldRoutes(old); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func replaceDefaultRoutes(from, to int) error {
+	routes, err := netlink.RouteListFiltered(netlink.FAMILY_V4, &netlink.Route{
+		Table: from,
+	}, netlink.RT_FILTER_TABLE)
+	if err != nil {
+		return err
+	}
+
+	for _, r := range routes {
+		route := r
+		route.Table = to
+		if err := netlink.RouteReplace(&route); err != nil {
+			return fmt.Errorf("unable to replace route under table ID: %w", err)
+		}
+	}
+
+	routes, _ = netlink.RouteListFiltered(netlink.FAMILY_V4, &netlink.Route{
+		Table: to,
+	}, netlink.RT_FILTER_TABLE)
+
+	return nil
+}
+
+func createNewRules(fromPrio, oldTable, toPrio, newTable int) error {
+	rules, err := findRulesByPriority(fromPrio)
+	if err != nil {
+		return err
+	}
+
+	candidates := make([]netlink.Rule, 0, len(rules))
+	for _, r := range rules {
+		if r.Table == oldTable {
+			candidates = append(candidates, r)
+		}
+	}
+
+	for _, r := range candidates {
+		rule := r
+		rule.Priority = toPrio
+		rule.Table = newTable
+		if err := netlink.RuleAdd(&rule); err != nil {
+			return fmt.Errorf("unable to add new rule: %w", err)
+		}
+	}
+
+	return nil
+}
+
+func deleteOldRules(prio int) error {
+	rules, err := findRulesByPriority(prio)
+	if err != nil {
+		return err
+	}
+
+	for _, r := range rules {
+		if err := netlink.RuleDel(&r); err != nil {
+			return fmt.Errorf("unable to delete old rules: %w", err)
+		}
+	}
+
+	return nil
+}
+
+func findRulesByPriority(prio int) ([]netlink.Rule, error) {
+	rules, err := netlink.RuleList(netlink.FAMILY_V4)
+	if err != nil {
+		return nil, err
+	}
+
+	candidates := make([]netlink.Rule, 0, len(rules))
+	for _, r := range rules {
+		if r.Priority == prio {
+			candidates = append(candidates, r)
+		}
+	}
+
+	return candidates, nil
+}
+
+func deleteOldRoutes(table int) error {
+	routes, err := netlink.RouteListFiltered(netlink.FAMILY_V4, &netlink.Route{
+		Table: table,
+	}, netlink.RT_FILTER_TABLE)
+	if err != nil {
+		return err
+	}
+
+	for _, r := range routes {
+		if err := netlink.RouteDel(&r); err != nil {
+			return fmt.Errorf("unable to delete old route: %w", err)
+		}
+	}
+
+	return nil
+}
+
+type ciliumNodeGetter interface {
+	Get(name string, opts v1.GetOptions) (*v2.CiliumNode, error)
+}
```

Comment:
Organizationally it seems a little odd to include k8s and CRD APIs from the `datapath/linux/routing package`. At first glance I wonder whether it'd be cleaner for resolving dependencies to expose a lower-level API instead, eg something that allows `GetInterfaceNumberByMac()`:

```suggestion
type InterfaceDB interface {
	GetInterfaceNumberByMAC(mac string) (int, error)
}
```